### PR TITLE
[inductor] Fold small-row expanded bmm to mm in post-grad

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185404

A dynamic linear on batch-first transposed activations can reach post-grad as
an expanded batched matmul:

    [T, B, K] @ expand([K, N], [T, K, N])

For small B this produces many small-row GEMMs. On CUDA fp16 this path can be
much slower than flattening the transposed activation to [T * B, K] and using
one mm. The issue shows up in AOTInductor autocast because the fp16 cast drops
requires_grad, so the existing matmul folding heuristic no longer takes the
requires_grad escape hatch that the fp32 graph happened to use.

Add a post-grad Inductor pattern for the specific expanded-bmm shape generated
by transposed linear. This keeps eager and general decomposition stride behavior
unchanged, addressing the concern from the abandoned decomposition-only PR. The
pattern is deliberately narrow: GPU fp16/bf16/fp32, batch-first transpose
layout, large T/K/N, and B <= 64. Dynamic dimensions are checked with guards so
cache reuse is stable: large batches keep the original bmm path, while small
batches compile to the mm path.

I considered a runtime torch.cond between bmm and mm, but tracing cond inside
this post-grad replacement currently nests torch.compile and trips Inductor's
compile event stack. Guarded specialization is the simpler established path and
keeps large-batch behavior on the original implementation.

Benchmark Results:
On NVIDIA B200, fp16, T=1000, K=2048, N=6144, max_autotune_gemm_backends=ATEN,
5 warmup iterations and 20 timed CUDA-event iterations:

Before:
- B=1: current bmm 0.0721 ms, manual folded mm 0.0724 ms
- B=2: current bmm 1.3912 ms, manual folded mm 0.0841 ms
- B=4: current bmm 1.4169 ms, manual folded mm 0.1524 ms

After:
- generated code uses mm for the small-batch guarded graph
- B=1: 0.0772 ms
- B=2: 0.0903 ms
- B=4: 0.1531 ms

Test Plan:
- python test/inductor/test_pattern_matcher.py -k test_transposed_linear_dynamic_batch_uses_guarded_mm
- python test/inductor/test_pattern_matcher.py -k test_bmm_to_mm
- lintrunner -a torch/_inductor/fx_passes/post_grad.py test/inductor/test_pattern_matcher.py
- git diff --check

Fixes #159346
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I added the missing release notes label and reran the failed CI jobs. The prior failures looked like infrastructure/resource issues: the inductor timm shard lost CUDA visibility, and the py3.14 serialization shards were SIGKILLed on the 2GB serialization test.

**Comment by eellison** [comment]
cc @yushangdi

**Review by eellison — COMMENTED** [review]
Looks good, but, can we do a bit more benchmarking

**Comment by jansel** [comment]
My agent says I addressed the benchmarking ask by running a wider CUDA sweep on a local branch with this PR commit replayed onto the build-matched main. Shape was the issue-sized transposed linear pattern, `T=1000, K=2048, N=6144`, `max_autotune_gemm_backends=ATEN`, dynamic compile, median of 3 x 20 CUDA-event iterations after warmup. I compared the PR's compiled `F.linear(x.transpose(0, 1), w, b)` against an explicit folded-mm implementation and checked correctness plus output strides against eager.

```
dtype=float16
B=1   kind=other current=0.0677 ms folded_mm=0.0756 ms ratio=0.90x
B=2   kind=mm    current=0.0839 ms folded_mm=0.1082 ms ratio=0.78x
B=4   kind=mm    current=0.1531 ms folded_mm=0.1532 ms ratio=1.00x
B=8   kind=mm    current=0.3117 ms folded_mm=0.3121 ms ratio=1.00x
B=16  kind=mm    current=0.6108 ms folded_mm=0.6161 ms ratio=0.99x
B=32  kind=mm    current=1.2795 ms folded_mm=1.2214 ms ratio=1.05x
B=64  kind=mm    current=2.5720 ms folded_mm=2.6220 ms ratio=0.98x
B=128 kind=bmm   current=4.6951 ms folded_mm=5.0942 ms ratio=0.92x

dtype=bfloat16
B=1   kind=other current=0.0712 ms folded_mm=0.0712 ms ratio=1.00x
B=2   kind=mm    current=0.0834 ms folded_mm=0.0846 ms ratio=0.99x
B=4   kind=mm    current=0.1514 ms folded_mm=0.1515 ms ratio=1.00x
B=8   kind=mm    current=0.3087 ms folded_mm=0.3079 ms ratio=1.00x
B=16  kind=mm    current=0.5996 ms folded_mm=0.5991 ms ratio=1.00x
B=32  kind=mm    current=1.1846 ms folded_mm=1.2120 ms ratio=0.98x
B=64  kind=mm    current=2.4594 ms folded_mm=2.4517 ms ratio=1.00x
B=128 kind=bmm   current=4.4696 ms folded_mm=4.9786 ms ratio=0.90x

dtype=float32
B=1   kind=other current=0.5159 ms folded_mm=0.5098 ms ratio=1.01x
B=2   kind=mm    current=0.9508 ms folded_mm=0.9496 ms ratio=1.00x
B=4   kind=mm    current=1.7921 ms folded_mm=1.7922 ms ratio=1.00x
B=8   kind=mm    current=3.5038 ms folded_mm=3.5045 ms ratio=1.00x
B=16  kind=mm    current=6.6748 ms folded_mm=6.6740 ms ratio=1.00x
B=32  kind=mm    current=13.0308 ms folded_mm=13.0299 ms ratio=1.00x
B=64  kind=mm    current=25.9250 ms folded_mm=25.9250 ms ratio=1.00x
B=128 kind=bmm   current=53.5029 ms folded_mm=51.6336 ms ratio=1.04x
```

This gives the intended small-batch fp16/bf16 path without a measurable penalty versus the explicit folded-mm baseline, and the guard still avoids folding at `B=128`.
